timing/timing.go

```diff
@@ -0,0 +1,172 @@
+package timing
+
+import "time"
+
+// Controls whether library calls should be mocked, or whether we should use the standard Go time library.
+// If we're in Mock Mode, then time does not pass as normal, but only progresses when Elapse is called.
+// False by default, indicating that we just call through to standard Go functions.
+var MockMode bool = false
+var currentTimeMock time.Time = time.Unix(0, 0)
+var mockTimers []*mockTimer = make([]*mockTimer, 0)
+
+// Interface over Golang's built-in Timers, allowing them to be swapped out for mocked timers.
+type Timer interface {
+	// Returns a channel which sends the current time immediately when the timer expires.
+	// Equivalent to time.Timer.C; however, we have to use a method here instead of a member since this is an interface.
+	C() <-chan time.Time
+
+	// Resets the timer such that it will expire in duration 'd' after the current time.
+	// Returns true if the timer had been active, and false if it had expired or been stopped.
+	Reset(d time.Duration) bool
+
+	// Stops the timer, preventing it from firing.
+	// Returns true if the timer had been active, and false if it had expired or been stopped.
+	Stop() bool
+}
+
+// Implementation of Timer that just wraps time.Timer.
+type realTimer struct {
+	*time.Timer
+}
+
+func (t *realTimer) C() <-chan time.Time {
+	return t.Timer.C
+}
+
+func (t *realTimer) Reset(d time.Duration) bool {
+	return t.Timer.Reset(d)
+}
+
+func (t *realTimer) Stop() bool {
+	return t.Timer.Stop()
+}
+
+// Implementation of Timer that mocks time.Timer, firing when the total elapsed time (as controlled by Elapse)
+// exceeds the duration specified when the timer was constructed.
+type mockTimer struct {
+	EndTime time.Time
+	Chan    chan time.Time
+}
+
+func (t *mockTimer) C() <-chan time.Time {
+	return t.Chan
+}
+
+func (t *mockTimer) Reset(d time.Duration) bool {
+	wasActive := removeMockTimer(t)
+
+	t.EndTime = currentTimeMock.Add(d)
+	if d == 0 {
+		t.Chan <- currentTimeMock
+	}
+
+	mockTimers = append(mockTimers, t)
```

Comment:
Is this correct?  Presumably we only want to append the timer to mockTimers in the d != 0 case?
Otherwise it'll proc here, and again at the next call to Elapse.